sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1167,6 +1192,73 @@ Exceptions:
   `property::graph::updatable` was not set when the executable graph was created.
 * Throws with error code `invalid` if any node in `nodes` is not part of the
   graph.
+
+|
+[source, c++]
+----
+void
+update(const command_graph<graph_state::modifiable>& graph);
+----
+
+|Updates the executable graph node inputs & outputs from a topologically
+identical modifiable graph. A topologically identical graph is one with the
+same structure of nodes and edges. Additionally, the nodes and edges must have
+been added in the same order to both graphs.
+
+Equivalent nodes in topologically identical graphs each have the
+same command. There is the additional limitation that to update an executable
+graph, every node in the graph must be a kernel command.
+
+During an update, the only characteristics that can differ between two
+topologically identical graphs are the arguments and the execution ND-Range of
+the kernel nodes. For example, the graph may capture different values for the
+USM pointers or accessors used in the graph. It is these kernels arguments in
+`graph` that constitute the inputs & outputs to update to.
+
+It is not an error to update an executable graph such that all the arguments and
+ND-Range of kernels in `graph` are identical to the arguments of the executable
+graph prior to the update.
+
+Differences in the following characteristics between two graphs during an
+update results in undefined behavior:
+
+* Modifying the {sycl-kernel-function}[kernel function] of a kernel node.
+* The order that nodes and edges were added to the graph.
+
+The implementation may perform a blocking wait during this call on
+any in-flight executions of that same graph if required by the backend.
+
+Constraints:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::executable`.
+
+Parameters:
+
+* `graph` - Modifiable graph object to update graph node inputs & outputs with.
+  This graph must have the same topology as the original graph used on
+  executable graph creation.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the topology of `graph` is
+  not the same as the existing graph topology in the following ways:
+  
+** The number of nodes in the graph does not match.
+** The number of edges between nodes does not match.
+
+
+:handler-copy-functions: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.members.handler.copy
```

Comment:
This is described already in the main description for this function above. I don't think it makes sense to add it here because this section is explicitly for exceptions which can be thrown, in keeping with general SYCL specification style.

Reply:
> Specifically, in GROMACS, we have such a case when multiple host threads are submitting tasks to the same graph, thus the order of the submitted operations is different despite the same graph topology; CUDA cannot handle an update either, but at least it raises an error on such topology changes so we can safely try the update (and if it fails, finalize the graph anew).

I am surprised that CUDA is able to reliably identify topology errors like that to make that approach safe. It was my undestanding that they can only reliably detect changes if similar guarantees are made so that they can reliably pair nodes between graphs. 

Some of those requirements to deterministically pair nodes between the graph being updated and the source graph can be seen in their docs [here](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#whole-graph-update). They do not refer to it as UB but to me it seems equivalent: if you do not respect those requirements then the errors returned in update are not reliable and could lead to incorrect behaviour.

Reply:
Here they have [a list of cases when they return an error](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1g34c8e1e6118aa68fc715250dfd93c1c4):

> `cudaGraphExecUpdate` sets the result member of `resultInfo` to `cudaGraphExecUpdateErrorTopologyChanged` under the following conditions:
> -   The count of nodes directly in `hGraphExec` and `hGraph` differ, in which case `resultInfo->errorNode` is set to NULL.
> -   `hGraph` has more exit nodes than `hGraph`, in which case `resultInfo->errorNode` is set to one of the exit nodes in `hGraph`.
> -  A node in `hGraph` has a different number of dependencies than the node from `hGraphExec` it is paired with, in which case `resultInfo->errorNode` is set to the node from `hGraph`.
> -  A node in `hGraph` has a dependency that does not match with the corresponding dependency of the paired node from `hGraphExec`. `resultInfo->errorNode` will be set to the node from `hGraph`. `resultInfo->errorFromNode` will be set to the mismatched dependency. The dependencies are paired based on edge order and a dependency does not match when the nodes are already paired based on other edges examined in the graph.

So far, the current proposal has the first three checks, but lacks the last one.

There are also a bunch of other error codes (change of node type etc). To me, at least, there are no gaps between requirements and the list of errors.

> They do not refer to it as UB but to me it seems equivalent: if you do not respect those requirements then the errors returned in update are not reliable and could lead to incorrect behaviour.

This raises the question of what "incorrect behavior" means. I am not sure how the graph update works internally, but if the there is a guarantee that, after a successful update the graph will behave the same as a newly-recorded one, I don't see any potential problems. If the current SYCL Graph code does that, then not raising an error should be fine; it is not an analysis tool to check whether user code follows the spec :)

Reply:
> So far, the current proposal has the first three checks, but lacks the last one.
>
>There are also a bunch of other error codes (change of node type etc). To me, at least, there are no gaps between requirements and the list of errors.

Perhaps I am misunderstanding something but the last error check you listed seems predicated on being able to deterministically pair edges, which is itself determined by the requirements in the link I provided. To my mind it seems that if you violate those requirements, and thus the pairing is no longer deterministic that this error cannot possibly be 100% reliable and may lead to false positives or negatives. So in your case of submitting from multiple threads, are you not potentially violating those requirements if API calls happen in a different order depending on exact thread timing? 

> This raises the question of what "incorrect behavior" means. I am not sure how the graph update works internally, but if the there is a guarantee that, after a successful update the graph will behave the same as a newly-recorded one, I don't see any potential problems.

If there is something which we say is an error and the user violates but which we cannot reliably 100% detect, to my mind if the update succeeds where it shouldn't that is not correct program behaviour AKA undefined behaviour.

Reply:
I feel like there is some mis-alignment of the underlying assumptions.

Here is how I see it:

- Why have the new "full graph update" API when one can call `finalize` on the "new" graph instead? Because updating a few nodes is faster than rebuilding the whole graph, but, for the API user, it is not always convenient to use per-node update. The goal is to have the "updated" graph behave the same as the "new" graph.
- Why require matching topology between the "old" and the "new" graphs? Because the "whole graph update" consists of a set of individual node updates, and we cannot cheaply change the topology, plus we need to match the nodes in the old and the new graphs to do this per-node update.
- Whether a single "old" node can be updated with the "new" node is clearly defined and can be reliably verified in the runtime.

If this is the case:

- if the matching cannot be constructed at all (e.g., different number of nodes/edges), the update will reliably fail;
- if the matching is constructed, but there is a "non-updateable" node pair (e.g., kernel and a memset, once those are supported), the update will reliably fail;
- if the matching is constructed and the order of operations is the same, we successfully update the minimal number of nodes that need to be updated;
- (here it gets fun) if the matching is constructed, but, due to the multithreading, we have a different order of nodes, but all the pairs are still "updateable" (e.g., in the example below, `NodeA` can be updated with `NodeC`, `NodeB` can be updated with `NodeD`, and vice-versa), then we do more updates then necessary (e.g., if only parameters to NodeB changed, we still update all four nodes), but the resulting update still succeeds, and the resulting "updated" graph will behave the same as the "new" graph.

While "updating four nodes instead of one" in the last case is not good, I think it is better than "UB, no diagnostics required".

old:

```
NodeA        NodeC
  |             |
NodeB        NodeD
```

new:

```
NodeC        NodeA
  |             |
NodeD        NodeB
```

Reply:
Thanks for the explanation I think I understand better where the differences lie here.

> (here it gets fun) if the matching is constructed, but, due to the multithreading, we have a different order of nodes, but all the pairs are still "updateable" (e.g., in the example below, NodeA can be updated with NodeC, NodeB can be updated with NodeD, and vice-versa), then we do more updates then necessary (e.g., if only parameters to NodeB changed, we still update all four nodes), but the resulting update still succeeds, and the resulting "updated" graph will behave the same as the "new" graph.

I think this is the main thing we disagreed on, and I think it was due to an incorrect assumption on my part. I thought that there could be a difference in behaviour between the updated graph and the new graph if this swap occurred as in your example but after some more consideration I think that is not the case as long as the kernel types match (which is now a requirement for matching topology in the spec here).

I've added exception wording for the case where dependencies of a node have a mismatch which should bring us in line with CUDA in terms of errors.